File: src/functions/event_api.py
"""Event API — 이벤트 수신, 저장, 조회.

POST /events: 이벤트 생성 (Cosmos DB 저장 + Idempotency)
GET /events/{event_id}: 이벤트 상세 조회
GET /events: 이벤트 목록 조회 (페이지네이션)

SPEC.md §7, §8.1, §8.2 참조.
"""

from __future__ import annotations

import json
import logging
from datetime import UTC, datetime
from typing import Any

import azure.functions as func
from azure.cosmos.exceptions import CosmosResourceExistsError, CosmosResourceNotFoundError

from src.models.events import (
    EventStatus,
    NotificationChannel,
    NotificationChannelType,
    NotificationEvent,
    NotificationStatus,
)
from src.services.cosmos_client import get_events_container
from src.shared.config import load_settings
from src.shared.correlation import clear_context, generate_correlation_id, set_correlation_id, set_log_context
from src.shared.errors import ValidationError as AppValidationError
from src.shared.logger import log_with_context
from src.shared.validator import validate_create_event

logger = logging.getLogger(__name__)

bp = func.Blueprint()  # type: ignore[no-untyped-call]

_settings = None


def _get_settings() -> Any:
    global _settings
    if _settings is None:
        _settings = load_settings()
    return _settings


def _build_notifications(
    channels: list[NotificationChannelType],
    settings: Any,
) -> list[NotificationChannel]:
    """channels 목록에서 notifications 초기 배열을 생성한다."""
    notifications = []
    for ch in channels:
        if ch == NotificationChannelType.EMAIL:
            provider = settings.NOTIFICATION_EMAIL_PROVIDER
        elif ch == NotificationChannelType.SMS:
            provider = settings.NOTIFICATION_SMS_PROVIDER
        else:
            provider = "webhook"
        notifications.append(
            NotificationChannel(
                channel=ch,
                provider=provider,
                status=NotificationStatus.PENDING,
            )
        )
    return notifications


def _json_response(body: dict[str, Any], status_code: int = 200) -> func.HttpResponse:
    """JSON HttpResponse를 생성한다."""
    return func.HttpResponse(
        body=json.dumps(body, ensure_ascii=False, default=str),
        status_code=status_code,
        mimetype="application/json",
    )


def _error_response(
    error_code: str,
    message: str,
    status_code: int,
    details: list[Any] | None = None,
) -> func.HttpResponse:
    """에러 JSON 응답을 생성한다."""
    return _json_response(
        {"error": error_code, "message": message, "details": details or []},
        status_code=status_code,
    )


@bp.route(route="events", methods=["POST"], auth_level=func.AuthLevel.ANONYMOUS)
async def post_events(req: func.HttpRequest) -> func.HttpResponse:
    """POST /events — 이벤트 생성.

    Pydantic 입력 검증 → correlation_id 생성 → Cosmos DB 저장.
    409 Conflict 시 기존 문서 조회 후 200 반환.
    """
    clear_context()
    correlation_id = generate_correlation_id()
    set_correlation_id(correlation_id)

    try:
        body = req.get_json()
    except ValueError:
        return _error_response("VALIDATION_ERROR", "Invalid JSON body", 400)

    # 입력 검증
    try:
        validated = validate_create_event(body)
    except AppValidationError as e:
        return _json_response(e.to_dict(), status_code=400)

    set_log_context(event_id=validated.id, clinic_id=validated.clinic_id)

    settings = _get_settings()
    now = datetime.now(tz=UTC)

    # NotificationEvent 문서 생성
    event_doc = NotificationEvent(
        id=validated.id,
        clinic_id=validated.clinic_id,
        status=EventStatus.QUEUED,
        event_type=validated.event_type,
        patient_id=validated.patient_id,
        channels=validated.channels,
        correlation_id=correlation_id,
        notifications=_build_notifications(validated.channels, settings),
        created_at=now,
        updated_at=now,
    )

    container = get_events_container(settings)

    try:
        doc_dict = event_doc.model_dump(by_alias=True)
        await container.create_item(body=doc_dict)

        log_with_context(logger, logging.INFO, "이벤트 생성 완료", status="queued")

        return _json_response(
            {"event_id": validated.id, "status": "queued", "correlation_id": correlation_id},
            status_code=201,
        )

    except CosmosResourceExistsError:
        # Idempotency: 기존 문서 조회 후 200 반환
        log_with_context(logger, logging.INFO, "중복 이벤트 요청", event_id=validated.id)

        try:
            existing = await container.read_item(item=validated.id, partition_key=validated.clinic_id)
            return _json_response(
                {
                    "event_id": existing["id"],
                    "status": existing.get("status", "queued"),
                    "correlation_id": existing.get("correlation_id", correlation_id),
                    "message": "Event already exists",
                },
                status_code=200,
            )
        except CosmosResourceNotFoundError:
            return _error_response("NOT_FOUND", "Event not found", 404)

    except Exception:
        logger.exception("이벤트 생성 실패")
        return _error_response("INTERNAL_ERROR", "Internal server error", 500)


@bp.route(route="events/{event_id}", methods=["GET"], auth_level=func.AuthLevel.ANONYMOUS)
async def get_event_by_id(req: func.HttpRequest) -> func.HttpResponse:
    """GET /events/{event_id} — 이벤트 상세 조회.

    clinic_id 쿼리 파라미터 필수.
    """
    clear_context()
    event_id = req.route_params.get("event_id", "")
    clinic_id = req.params.get("clinic_id")

    if not clinic_id:
        return _error_response("VALIDATION_ERROR", "clinic_id query parameter is required", 400)

    set_log_context(event_id=event_id, clinic_id=clinic_id)

    settings = _get_settings()
    container = get_events_container(settings)

    try:
        doc = await container.read_item(item=event_id, partition_key=clinic_id)
        # _outbox_status 내부 필드 제거
        doc.pop("_outbox_status", None)
        # Cosmos DB 메타 필드 제거
        for key in ("_rid", "_self", "_etag", "_attachments", "_ts"):
            doc.pop(key, None)
        return _json_response(doc)

    except CosmosResourceNotFoundError:
        return _error_response("NOT_FOUND", f"Event {event_id} not found", 404)

    except Exception:
        logger.exception("이벤트 조회 실패")
        return _error_response("INTERNAL_ERROR", "Internal server error", 500)


@bp.route(route="events", methods=["GET"], auth_level=func.AuthLevel.ANONYMOUS)
async def get_events(req: func.HttpRequest) -> func.HttpResponse:
    """GET /events — 이벤트 목록 조회.

    clinic_id 필수, status/event_type 필터, continuation_token 페이지네이션.
    """
    clear_context()
    clinic_id = req.params.get("clinic_id")

    if not clinic_id:
        return _error_response("VALIDATION_ERROR", "clinic_id query parameter is required", 400)

    set_log_context(clinic_id=clinic_id)

    status_filter = req.params.get("status")
    event_type_filter = req.params.get("event_type")
    continuation_token = req.params.get("continuation_token")
    page_size = min(int(req.params.get("page_size", "20")), 100)

    settings = _get_settings()
    container = get_events_container(settings)

    # 쿼리 빌드
    conditions = ["c.clinic_id = @clinic_id"]
    parameters: list[dict[str, object]] = [{"name": "@clinic_id", "value": clinic_id}]

    if status_filter:
        conditions.append("c.status = @status")
        parameters.append({"name": "@status", "value": status_filter})

    if event_type_filter:
        conditions.append("c.event_type = @event_type")
        parameters.append({"name": "@event_type", "value": event_type_filter})

    select_fields = (
        "c.id, c.clinic_id, c.status, c.event_type, c.patient_id, "
        "c.channels, c.correlation_id, c.created_at, c.updated_at"
    )
    where_clause = " AND ".join(conditions)
    query = f"SELECT {select_fields} FROM c WHERE {where_clause} ORDER BY c.created_at DESC"

    try:
        items: list[dict[str, Any]] = []
        new_continuation_token = None

        query_iterable = container.query_items(
            query=query,
            parameters=parameters,
            partition_key=clinic_id,
            max_item_count=page_size,
        )

        # 페이지네이션 처리
        pager = query_iterable.by_page(continuation_token)
        async for page in pager:
            async for item in page:
                items.append(item)
            new_continuation_token = pager.continuation_token  # type: ignore[attr-defined]
            break  # 첫 페이지만

        return _json_response(
            {
                "items": items,
                "continuation_token": new_continuation_token,
            }
        )

    except Exception:
        logger.exception("이벤트 목록 조회 실패")
        return _error_response("INTERNAL_ERROR", "Internal server error", 500)

File: src/functions/event-api.ts
/**
 * Event API — 이벤트 수신, 저장, 조회.
 *
 * POST /events: 이벤트 생성 (Cosmos DB 저장 + Idempotency)
 * GET /events/{event_id}: 이벤트 상세 조회
 * GET /events: 이벤트 목록 조회 (페이지네이션)
 *
 * SPEC.md §7, §8.1, §8.2 참조.
 */

import { app, type HttpRequest, type HttpResponseInit, type InvocationContext } from '@azure/functions';
import type { NotificationChannel, NotificationChannelType, NotificationEvent } from '../models/events';
import { EventStatus, NotificationStatus, OutboxStatus } from '../models/events';
import type { Settings } from '../shared/config';
import { getSettings } from '../shared/config';
import { clearContext, generateCorrelationId, runWithContext, setCorrelationId, setLogContext } from '../shared/correlation';
import { ValidationError } from '../shared/errors';
import { getLogger, logWithContext } from '../shared/logger';
import { validateCreateEvent } from '../shared/validator';
import { getEventsContainer } from '../services/cosmos-client';

const logger = getLogger('event-api');

let _settings: Settings | null = null;
function _getSettings(): Settings {
  if (!_settings) _settings = getSettings();
  return _settings;
}

/** 테스트용 settings 오버라이드 */
export function _setSettingsForTest(settings: Settings): void {
  _settings = settings;
}

function buildNotifications(
  channels: NotificationChannelType[],
  settings: Settings,
): NotificationChannel[] {
  return channels.map((ch) => {
    let provider: string;
    if (ch === 'email') provider = settings.NOTIFICATION_EMAIL_PROVIDER;
    else if (ch === 'sms') provider = settings.NOTIFICATION_SMS_PROVIDER;
    else provider = 'webhook';

    return {
      channel: ch,
      provider,
      status: NotificationStatus.PENDING,
      sent_at: null,
      retry_count: 0,
      last_error: null,
    };
  });
}

function jsonResponse(body: Record<string, unknown>, status: number = 200): HttpResponseInit {
  return {
    status,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  };
}

function errorResponse(
  errorCode: string,
  message: string,
  status: number,
  details: unknown[] = [],
): HttpResponseInit {
  return jsonResponse({ error: errorCode, message, details }, status);
}

export async function postEvents(
  req: HttpRequest,
  _context: InvocationContext,
): Promise<HttpResponseInit> {
  return runWithContext(async () => {
    clearContext();
    const correlationId = generateCorrelationId();
    setCorrelationId(correlationId);

    let body: unknown;
    try {
      body = await req.json();
    } catch {
      return errorResponse('VALIDATION_ERROR', 'Invalid JSON body', 400);
    }

    let validated;
    try {
      validated = validateCreateEvent(body);
    } catch (err: unknown) {
      if (err instanceof ValidationError) {
        return jsonResponse(err.toDict() as Record<string, unknown>, 400);
      }
      return errorResponse('VALIDATION_ERROR', 'Invalid request body', 400);
    }

    setLogContext({ event_id: validated.id, clinic_id: validated.clinic_id });

    const settings = _getSettings();
    const now = new Date().toISOString();

    const eventDoc: NotificationEvent = {
      id: validated.id,
      clinic_id: validated.clinic_id,
      status: EventStatus.QUEUED,
      event_type: validated.event_type as NotificationEvent['event_type'],
      patient_id: validated.patient_id,
      channels: validated.channels as NotificationChannelType[],
      correlation_id: correlationId,
      notifications: buildNotifications(validated.channels as NotificationChannelType[], settings),
      created_at: now,
      updated_at: now,
      _outbox_status: OutboxStatus.PENDING,
    };

    const container = getEventsContainer(settings);

    try {
      await container.items.create(eventDoc);
      logWithContext(logger, 'INFO', '이벤트 생성 완료', { status: 'queued' });

      return jsonResponse(
        { event_id: validated.id, status: 'queued', correlation_id: correlationId },
        201,
      );
    } catch (err: unknown) {
      // Idempotency: 409 Conflict → 기존 문서 반환
      if (typeof err === 'object' && err !== null && 'code' in err && (err as { code: number }).code === 409) {
        logWithContext(logger, 'INFO', '중복 이벤트 요청', { event_id: validated.id });
        try {
          const { resource: existing } = await container.item(validated.id, validated.clinic_id).read();
          if (existing) {
            return jsonResponse({
              event_id: existing.id as string,
              status: (existing.status as string) ?? 'queued',
              correlation_id: (existing.correlation_id as string) ?? correlationId,
              message: 'Event already exists',
            }, 200);
          }
        } catch {
          return errorResponse('NOT_FOUND', 'Event not found', 404);
        }
      }
      logger.error('이벤트 생성 실패', { error: String(err) });
      return errorResponse('INTERNAL_ERROR', 'Internal server error', 500);
    }
  });
}

export async function getEventById(
  req: HttpRequest,
  _context: InvocationContext,
): Promise<HttpResponseInit> {
  return runWithContext(async () => {
    clearContext();
    const eventId = req.params.event_id ?? '';
    const clinicId = req.query.get('clinic_id');

    setLogContext({ event_id: eventId, clinic_id: clinicId ?? '' });

    const settings = _getSettings();
    const container = getEventsContainer(settings);

    try {
      let doc: Record<string, unknown> | undefined;

      if (clinicId) {
        // 파티션 키가 있으면 point read (빠름)
        const { resource } = await container.item(eventId, clinicId).read();
        doc = resource as Record<string, unknown> | undefined;
      } else {
        // 파티션 키 없으면 cross-partition query
        const { resources } = await container.items.query({
          query: 'SELECT * FROM c WHERE c.id = @id',
          parameters: [{ name: '@id', value: eventId }],
        }).fetchAll();
        doc = resources[0] as Record<string, unknown> | undefined;
      }

      if (!doc) return errorResponse('NOT_FOUND', `Event ${eventId} not found`, 404);

      // 내부 필드 제거
      const result = { ...doc };
      delete result._outbox_status;
      for (const key of ['_rid', '_self', '_etag', '_attachments', '_ts']) {
        delete result[key];
      }

      return jsonResponse(result);
    } catch (err: unknown) {
      if (typeof err === 'object' && err !== null && 'code' in err && (err as { code: number }).code === 404) {
        return errorResponse('NOT_FOUND', `Event ${eventId} not found`, 404);
      }
      logger.error('이벤트 조회 실패', { error: String(err) });
      return errorResponse('INTERNAL_ERROR', 'Internal server error', 500);
    }
  });
}

export async function getEvents(
  req: HttpRequest,
  _context: InvocationContext,
): Promise<HttpResponseInit> {
  return runWithContext(async () => {
    clearContext();
    const clinicId = req.query.get('clinic_id');

    if (!clinicId) {
      return errorResponse('VALIDATION_ERROR', 'clinic_id query parameter is required', 400);
    }

    setLogContext({ clinic_id: clinicId });

    const statusFilter = req.query.get('status');
    const eventTypeFilter = req.query.get('event_type');
    const continuationToken = req.query.get('continuation_token') ?? undefined;
    const pageSize = Math.min(parseInt(req.query.get('page_size') ?? '20', 10) || 20, 100);

    const settings = _getSettings();
    const container = getEventsContainer(settings);

    const conditions = ['c.clinic_id = @clinic_id'];
    const parameters: { name: string; value: string }[] = [
      { name: '@clinic_id', value: clinicId },
    ];

    if (statusFilter) {
      conditions.push('c.status = @status');
      parameters.push({ name: '@status', value: statusFilter });
    }

    if (eventTypeFilter) {
      conditions.push('c.event_type = @event_type');
      parameters.push({ name: '@event_type', value: eventTypeFilter });
    }

    const selectFields =
      'c.id, c.clinic_id, c.status, c.event_type, c.patient_id, ' +
      'c.channels, c.correlation_id, c.created_at, c.updated_at';
    const whereClause = conditions.join(' AND ');
    const query = `SELECT ${selectFields} FROM c WHERE ${whereClause} ORDER BY c.created_at DESC`;

    try {
      const queryIterator = container.items.query(
        { query, parameters },
        {
          partitionKey: clinicId,
          maxItemCount: pageSize,
          continuationToken,
        },
      );

      const response = await queryIterator.fetchNext();
      const items = response.resources ?? [];
      const newContinuationToken = response.continuationToken ?? null;

      return jsonResponse({ items, continuation_token: newContinuationToken });
    } catch (err: unknown) {
      logger.error('이벤트 목록 조회 실패', { error: String(err) });
      return errorResponse('INTERNAL_ERROR', 'Internal server error', 500);
    }
  });
}

// Azure Functions v4 HTTP 등록
app.http('postEvents', {
  methods: ['POST'],
  route: 'events',
  authLevel: 'anonymous',
  handler: postEvents,
});

app.http('getEventById', {
  methods: ['GET'],
  route: 'events/{event_id}',
  authLevel: 'anonymous',
  handler: getEventById,
});

app.http('getEvents', {
  methods: ['GET'],
  route: 'events',
  authLevel: 'anonymous',
  handler: getEvents,
});
